Feat/order payment flow

miooo0o opened the pull request:
# Order-Payment Flow
The `CheckoutController` is responsible for the overall order and payment process. Refund features will be added here in the future.

To clarify, I created several entities and their corresponding repos and services, and then merged them in the `CheckoutController `to implement their usage. **I touched some entities during this process (sorry)**, so there may be conflicts (`Address`,` Member`, `Order`, `PackageOption`, etc.). I explained most of you people in advance, but I'm writing this down for sure!

While I was implementing this, I merged in the parts you reviewed separately @VittorioDeMarzi  @pebencze, but there may be some changes.

Thank you!

## Description

Enhance 2 Endpoints 

### 1. **POST** `/api/checkout/start` at `CheckoutController`
Orders and payments are bundled together as a Checkout session (such as `PaymentIntent`).
During the checkout process:

- Creates `Order` from `Cart` and reduces stock
- Creates `PaymentIntent` via `PaymentService`  
- Returns payment details (clientSecret, orderId, amount)
- **Runs in single transaction** - auto rollback on failure

```kotlin
@PostMapping("/start")
fun start(
    @LoginMember member: MemberDto,
    @Valid @RequestBody request: StartCheckoutRequest,
): ResponseEntity<StartCheckoutResponse> {
    val response = checkoutService.startCheckout(member, request)
    return ResponseEntity.ok(response)
}
```

### 2. **POST** `/api/checkout/finalize` at `CheckoutController`

Payment confirmation is handled through the finalize process.
During finalization:
- Confirms PaymentIntent with payment provider
- Marks Order as paid/failed/cancelled based on payment status
- Rollbacks stock if payment fails
- **Runs in single transaction** - ensures data consistency

```kotlin
@PostMapping("/finalize")
fun finalize(
    @LoginMember member: MemberDto,
    @Valid @RequestBody request: FinalizePaymentRequest,
): ResponseEntity<FinalizePaymentResponse> {
    val response = checkoutService.finalizeCheckout(member, request)
    return ResponseEntity.ok(response)
}
```

## Note & Questions For Reviewers

I need to write a PR for a large PR that has already grown in size, and I'm writing it for the current state. If there are any parts that need fixing, please let me know!

### Todo

- [ ] Set maximum time limits for checkout processes, preventing writes from running indefinitely or maintaining meaningless lock states.
- [ ] Better exception handler for Stripe operations
- [ ] E2E test or basic test with Postman
- [ ] Payment Status / Order Status (Integrate or clarify usage. Currently being used interchangeably)
- [ ] Check spelling of "canceled" in Stripe

---

Some items can be converted to new tickets

## Type of change

- [x] Bug fix (change which fixes an issue): non nullable id
- [x] New feature (change which adds functionality)
- [x] Refactoring (change of existing codebase)
- [ ] Chore (build process or tool changes)
- [ ] Style (formatting code)
- [ ] Docs (updating or adding a readme, commenting code)
- [ ] Test (adding new tests)
- [ ] CI (changes to CI related scripts or files)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Test

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.
Please choose what type of test you have used:
- Slice: OrderService, PaymentService
- Unit:

Perhaps still need more tests. Please let me know any case which you want me to try!

## Resources & Documentation

Projects:
- [](https://github.com/users/VittorioDeMarzi/projects/1/views/3?pane=issue&itemId=124742166&issue=VittorioDeMarzi%7Chero-beans%7C81) #81 

## Checklist:

- [x] I have performed a self-review of my code
- [x] My code is readable
- [x] I have run `ktlintFormat`
- [x] No wildcard imports or dead code
- [x] No sensitive data in logs or commits
- [x] The application runs
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (sorry, not many)
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the documentation
- [ ] SOLID principles and DRY are applied -- not sure

miooo0o committed:
feat(stripe-client, properties): add stripe client with basic DTOs

Flow:
* Implementation incomplete
- `/start`: creates order + payment intent (confirm=false)
- Frontend handles response, receives new request from Stripe
- New request returns via FinalizePaymentRequest to `/finalize`

miooo0o committed:
feat(stripe-client, properties): add stripe client with basic DTOs

Flow:
* Implementation incomplete
- `/start`: creates order + payment intent (confirm=false)
- Frontend handles response, receives new request from Stripe
- New request returns via FinalizePaymentRequest to `/finalize`

miooo0o committed:
feat(repository): add Order, OrderItem, PackageOption repositories

miooo0o committed:
feat(repository): add Order, OrderItem, PackageOption repositories

miooo0o committed:
feat(checkout-service, payment): integrate Payment entity from codebase and implement Stripe services

- Add Payment entity from existing codebase
- Implement Stripe payment service for checkout flow
- Add checkout service integration

miooo0o committed:
feat(checkout-service, payment): integrate Payment entity from codebase and implement Stripe services

- Add Payment entity from existing codebase
- Implement Stripe payment service for checkout flow
- Add checkout service integration

miooo0o committed:
refactor(package-option): make quantity methods return self

- Update decreaseQuantity/increaseQuantity to return PackageOption

miooo0o committed:
refactor(package-option): make quantity methods return self

- Update decreaseQuantity/increaseQuantity to return PackageOption

miooo0o committed:
refactor(entity, enums): detailed changes to entities and DTOs before a Payment process

- Delegate Order creation process to service layer
- Add ShippingFee to ShippingMethod enum (still using magic numbers)

miooo0o committed:
refactor(entity, enums): detailed changes to entities and DTOs before a Payment process

- Delegate Order creation process to service layer
- Add ShippingFee to ShippingMethod enum (still using magic numbers)

miooo0o committed:
refactor(checkout-service, order-service): separate "processOrderWithStockReduction" to OrderService from CheckoutService

miooo0o committed:
refactor(checkout-service, order-service): separate "processOrderWithStockReduction" to OrderService from CheckoutService

miooo0o committed:
refactor(checkout, shipping-method): rename method to increase readability, remove unused parameter

- change method name `applyShippingPolicy` to `getShippingPolicy`
- FinalizePaymentRequest has orderId

miooo0o committed:
refactor(checkout, shipping-method): rename method to increase readability, remove unused parameter

- change method name `applyShippingPolicy` to `getShippingPolicy`
- FinalizePaymentRequest has orderId

miooo0o committed:
refactor(checkout, shipping-method): rename method to increase readability, remove unused parameter

- change method name `applyShippingPolicy` to `getShippingPolicy`
- FinalizePaymentRequest has orderId

miooo0o committed:
refactor(checkout, shipping-method): rename method to increase readability, remove unused parameter

- change method name `applyShippingPolicy` to `getShippingPolicy`
- FinalizePaymentRequest has orderId

miooo0o committed:
fix(order, payment): annotation alignment and remove mappedBy for proper relationship mapping

miooo0o committed:
fix(order, payment): annotation alignment and remove mappedBy for proper relationship mapping

miooo0o committed:
fix(stripe-properties): add default values for configuration properties

miooo0o committed:
fix(stripe-properties): add default values for configuration properties

miooo0o committed:
chore: cleanup code

miooo0o committed:
chore: cleanup code

miooo0o committed:
refactor(checkout-service): improve checkout service architecture

- Extract PaymentService for better separation of concerns
- Enhance exception handling and validation
- Improve code organization and maintainability
- Add proper error handling for payment operations

miooo0o committed:
refactor(checkout-service): improve checkout service architecture

- Extract PaymentService for better separation of concerns
- Enhance exception handling and validation
- Improve code organization and maintainability
- Add proper error handling for payment operations

miooo0o committed:
refactor(payment-service): require requested payment intent id to exist in Payment repository

miooo0o committed:
refactor(payment-service): require requested payment intent id to exist in Payment repository

miooo0o committed:
chore: add new todo comments, cleanup code

miooo0o committed:
chore: add new todo comments, cleanup code

miooo0o committed:
test: add new tests for OrderService and PaymentService

miooo0o committed:
test: add new tests for OrderService and PaymentService

miooo0o committed:
Merge remote-tracking branch 'origin/feat/order-payment-payment' into feat/order-payment-flow

miooo0o committed:
Merge remote-tracking branch 'origin/feat/order-payment-payment' into feat/order-payment-flow

- Renmae "feat/order-payment-payment" to "feat/order-payment-flow"

miooo0o committed:
Merge remote-tracking branch 'origin/feat/order-payment-payment' into feat/order-payment-flow

miooo0o committed:
refactor(payment-service): change confirmPaymentIntent to take paymentIntentId directly instead of from request

miooo0o committed:
feat(order): implement status management and validation logic

- Implement private status field with controlled access
- Add business methods for status transitions
- Validate status changes (PENDING -> PAID -> SHIPPED, etc.)
- Track shipping and delivery timestamps

miooo0o committed:
refactor(package-option): enhance inventory quantity validation logic

- Improve quantity range validation in increase/decrease methods
- Add proper bounds checking (MIN_QUANTITY to MAX_QUANTITY)
- Enhance error messages for better debugging
- Refactor checkAvailabilityInStock method for clearer logic

miooo0o committed:
chore: add "@ActiveProfiles("test")" for tests, clean code, add new todo lists

miooo0o committed:
refactor(member): update controller to handle non-nullable member id

miooo0o committed:
chore: add new environment

miooo0o committed:
Merge branch 'dev' into feat/order-payment-flow

miooo0o committed:
fix: conflict resolve/merge in OptionJpaRepository to PackageOptionJpaRepository

VittorioDeMarzi committed:
chore: add secret key for stripe

VittorioDeMarzi committed:
refactor: fix different stripe checkout flow

miooo0o committed:
test(stripe-client): add a new test for stripe

miooo0o committed:
refactor(checkout-dto): rename to CheckoutStartRequest and CheckoutStartResponse

miooo0o committed:
fix(stripe-client): fixed an issue where confirm wasn't working.

- The URL was the problem.

miooo0o committed:
refactor(checkout-service): exception handling in finalizeCheckout to separate Order and Payment concerns

* Order-related exceptions no longer wrapped in PaymentResult
  - OrderNotFoundException, UnauthorizedAccessException thrown directly
  - Global exception handler maps to appropriate HTTP status codes
* Only Payment/Stripe exceptions handled as PaymentResult.Failure
  - Stock rollback only occurs on actual payment failures
  - Eliminates unnecessary stock rollback logic
* Improved exception classification system
  - Clear domain separation: Order vs Payment exceptions
  - Enhanced readability for logging and monitoring
  - Optimized HTTP status code mapping (404, 422, 500)

miooo0o committed:
refactor(order-service): apply consistent exception handling and remove direct repository access from PackOption

- Remove optionRepository from CheckoutService, use PackageOptionService
- Apply order/payment exception separation to both create/finalize
- Update tests to mock services instead of repositories
- Consistent service layer abstraction throughout checkout flow

miooo0o committed:
test(payment, order): update tests to match service layer changes

- Mock PackageOptionService instead of PackageOptionJpaRepository

miooo0o committed:
refactor(exception): set StripeProcessingException as sealed class from open class

- Consistency with other exception groups

miooo0o committed:
fix(exception): set StripeProcessingException as open class

miooo0o committed:
test: disable stripe integration test

miooo0o commented:
## Payment Flow Refactoring 

Hello. I have added tests and completed refactoring according to several requirements. The tests don't provide perfect coverage - they only check happy scenarios.

This flow is essential for future development processes, so even though it's incomplete, I was submitting this PR.
1. **Please make sure to report any necessary parts/changes/issues you discover**
2. Strpe test is muted due to secret-key issue.

---

## What intended to implement
* When logging, make it readable and clearly indicate what situation the error is about in advance
* Distinguish between client, server, and other eternal errors from Stripe  
* Also distinguish between errors occurring in order, payment, and stripe
* Make coupon application easier

These requirements are very important because I will start developing refunds from tomorrow...! To avoid "where did this issue originate?" in future merges, I thought I needed to refine the flow.

For this reason, I added...

### About CheckoutService

The most important part of this PR is actually the CheckoutService. Other parts are supplementary and were created to complete this feature.

```kotlin
@Transactional
fun startCheckout(
    memberDto: MemberDto,
    request: CheckoutStartRequest,
): CheckoutStartResponse {
    // --
    val cart = cartService.getCartForOrder(memberDto.id)
    val order = orderService.processOrderWithStockReduction(cart) // Lock options here
    // --  Create an Order and reduce stock by the ordered quantity
    return try {
        // integrate with Stripe API 
        val paymentIntent = paymentService.createPaymentIntent(request, order.totalAmount)

        // save the attempt ID and ordered content in Payment
        val payment = paymentService.createPayment(request, paymentIntent, order)

        CheckoutStartResponse(...)
    } catch (exception: Exception) {
        val error = mapToPaymentError(exception)
        throw exception
    }
}

@Transactional
fun finalizeCheckout(
    member: MemberDto,
    request: FinalizePaymentRequest,
): PaymentResult {
    // find validated Order. Order's status should be pending
    val order = orderService.getValidatedPendingOrder(request.orderId, member.id)
    return try {
        // Analyze the result Stripe API provided for the payment
        val paymentIntent = paymentService.confirmPaymentIntent(request.paymentIntentId)

        // set Order.status as Paid and save Order
        val status = updateOrderToPaid(order, paymentIntent)

        PaymentResult.Success(orderId = order.id, paymentStatus = status.name)
    } catch (exception: Exception) {
        // Lock options here
        handleCheckoutFinalizeFailure(order, exception)
          // 1. rollback
          // 2. set Order.status as Payment Fail and save Order
    }
}
```

These two transactions: 
1. Create an Order and reduce stock by the ordered quantity in advance (it will rollback if it fails anyway!), integrate with Stripe API to make the initial "payment attempt", and save the attempt ID and ordered content in Payment.

2. Analyze the result Stripe API provided for the payment, and if payment failed, manually rollback the quantity "as much as was attempted" and generate error messages. If successful, return success. If written more "specifically", we could examine the required stripe status in detail, but I'll finish it at a level that works adequately.

---

### Error handling

```kotlin
sealed class PaymentResult(
    val response: FinalizePaymentResponse,
) {
    class Success(orderId: Long?, paymentStatus: String?) : PaymentResult(FinalizePaymentResponse(orderId, paymentStatus))

    class Failure(orderId: Long?, val error: PaymentError) : PaymentResult(FinalizePaymentResponse(orderId))
}

data class PaymentError(
    val code: PaymentErrorCode,
    private val message: String? = null,
    val cause: Throwable? = null,
) {
    fun message(msg: String?): PaymentError = this.copy(message = msg)
}
```

This PaymentResult only stores the result of the final payment completion. It would be great if this gets added to the database someday, but we don't have time for that, so we'll pass on that part.

This stored result can be used for logging and also for checking status codes. It's not used extensively, only for checking why a payment failed (so it's not generated during the process of sending Payment attempts to Stripe. Why? It's unnecessary at the current stage and would make the code too large). Payments must send a response "even if they fail"...

```kotlin
 } catch (exception: Exception) {
        val error = mapToPaymentError(exception) // not use, but I save it for later @AnnLvu 
        throw exception
    }
```

### Added Exception Groups

```kotlin
sealed class PaymentException
sealed class OrderException  
sealed class StripeProcessingException
```

These three are used to identify "where the error occurred".

## Please think about this

```kotlin
class PaymentIntentNotFoundException(message: String, override val cause: Throwable? = null) : StripeProcessingException(message, cause)
```

I currently placed this under StripeProcessingException, but I'm not sure if that's good.

pebencze reviewed (changes requested):
Hey Mina! Thanks for your hard work! I have added some comments, maybe 2-3 require changes, the others are just suggestions and thoughts. :) Looks good overall!

VittorioDeMarzi committed:
refactor(order): handle order flow and exceptions

- add cart exceptions
- return sealed payment response after payment
- clear cart after payment success
- better message for exceptions
- parse client response

VittorioDeMarzi reviewed (commented):
I haven't left many comments since we refactored the code together. Here are the main points for handling the order flow and exceptions
- Add cart exceptions: Implement specific exceptions for cart-related errors.
- Return sealed payment response: Ensure a sealed response is returned after the payment process.
- Clear cart on success: The cart should be cleared upon successful payment.
- Improve exception messages: Provide more descriptive and clearer messages for exceptions.
- Parse client response: The response from the client needs to be parsed correctly.

miooo0o committed:
chore: correct spelling, remove unused comment

miooo0o committed:
feat: Add comprehensive E2E tests for checkout controller and fix BigDecimal conversion

- Add E2E test coverage for checkout flow including start/finalize payment

miooo0o committed:
Merge remote-tracking branch 'origin/dev' into feat/order-payment-flow

# Conflicts:
#	src/main/kotlin/techcourse/herobeans/configuration/WebMvcConfigurer.kt
#	src/main/kotlin/techcourse/herobeans/controller/CartController.kt
#	src/main/kotlin/techcourse/herobeans/exception/GlobalExceptionHandler.kt
#	src/main/kotlin/techcourse/herobeans/service/CartService.kt
#	src/main/resources/application-test.properties
#	src/main/resources/application.properties

miooo0o committed:
resolve: merge conflicts in checkout and coupon test setup

miooo0o committed:
refactor(checkout-controller): remove unused Valid annotation

miooo0o reviewed (commented):
Thank you for entrusting me with this part.

(Personally) Before starting this part, I had no intention of actively using Stripe's API. While **webhooks** that work asynchronously are cool, I vaguely thought they would be difficult to test. I also thought that if we were to use webhooks, we'd need to approach it in an **event-driven** way, and since I believed the right process was to "first implement and apply" rather than do it "right away," I proposed developing without using webhooks.

I have some regrets now, but at that time I couldn't even finish the mission (I couldn't implement orders at all even in my Mission), and I could only imagine within the limits of what I knew at the time, so this was the best approach. However, I think actual development is similar in some ways. When implementation is urgent, there are times when you have to implement with the capabilities you have. And then you apply new technologies. If we had more time, I'd want to use webhooks. But since refunds are urgent, I don't want to modify this currently working flow.

---

### Why We Chose Pessimistic Lock

We have a background setting that we created for our convenience (especially for decision-making purposes).
> We run a small-scale coffee beans online shop and targeting the Berlin area.

Initially, we won't have many items, our users will be "at most" 10-30 people, and daily order volume will be only 10%.
In this state, there won't be situations where multiple people "order at the same moment." `Pessimistic lock` is used for `Write` and during that time, the data "cannot be newly written."

It's slow but easy to handle, and safer for protecting data. There's no reason not to use pessimistic in our user pool.

- Currently, it's only used in "Payment," but ideally, we should add more.

### Why We Chose This Start-Finalize Flow

Since Stripe handles "payment processing" in the middle, we needed at least a Start-Finalize approach.

1. Delivering the user's payment attempt to Stripe
2. Checking Stripe's response, parsing it, and storing it

There was no need to expand this flow further. We're small-scale and didn't think the initial PR needed to be huge, so there was no need to update it further. But this flow itself could be refined further. If anything more is needed in the refund process, we can update it.

miooo0o commented on src/test/kotlin/techcourse/herobeans/client/StripeClientTest.kt:
```diff
@@ -0,0 +1,54 @@
+import techcourse.herobeans.dto.CheckoutStartRequest
+import java.math.BigDecimal
+@Disabled("Stripe integration test - requires valid API key")
+@SpringBootTest(
```
Yes. However, as of now, this key cannot be uploaded to public/git.
@VittorioDeMarzi I need help getting this applied via GitHub Actions.

miooo0o commented on src/test/kotlin/techcourse/herobeans/client/StripeClientTest.kt:
```diff
@@ -0,0 +1,54 @@
+        assertThat(actual.amount).isEqualTo(1000)
+    }
+
+    @Test
```
Oh! good idea! I'd definitely like to add more tests.
But the main priority was to verify the flow and ensure the happy path scenario works, so I didn't add any additional parts to this test. I think I can create a separate ticket to address this later. What are your thoughts?

If you have any scenarios in mind, let me know. I'll do unit tests and try end-to-end testing.

miooo0o commented on src/main/kotlin/techcourse/herobeans/config/StripeProperties.kt:
```diff
@@ -0,0 +1,10 @@
+
+import org.springframework.boot.context.properties.ConfigurationProperties
+
+// TODO: delete this default value
```
Oh! I forgot to remove this comment. Thank you for let me know

miooo0o commented on src/main/kotlin/techcourse/herobeans/controller/CheckoutController.kt:
```diff
@@ -0,0 +1,42 @@
+@RestController
+@RequestMapping("/api/checkout")
+@Validated
+class CheckoutController(
```
I couldn't do the documentation yesterday. I'll do it today. Sorry for the delay!

miooo0o commented on src/main/kotlin/techcourse/herobeans/entity/Payment.kt:
```diff
@@ -0,0 +1,36 @@
+    @Enumerated(EnumType.STRING)
+    var status: PaymentStatus = PaymentStatus.PENDING,
+    @CreationTimestamp
+    var createdAt: LocalDateTime? = null,
```
I talked about this with @AnnLvu last time, and there was definitely a reason.
I can't remember detail now. I'll look it up again. 

And @AnnLvu, if you remember let's me know

miooo0o commented on src/main/kotlin/techcourse/herobeans/dto/StartCheckoutRequest.kt:
```diff
@@ -0,0 +1,7 @@
+
+class StartCheckoutRequest(
+    val paymentMethod: String,
+    // TODO: implement class
```
That's still I am not sure.
I feel like: "There's no real value in changing it to an Enum, but it keeps bothering me not to change it :("
Any idea or prefer?

miooo0o commented on src/main/kotlin/techcourse/herobeans/config/StripeProperties.kt:
```diff
@@ -0,0 +1,10 @@
+
+import org.springframework.boot.context.properties.ConfigurationProperties
+
+// TODO: delete this default value
```
When we were refactoring this together, there was a default. I removed this comment! thank you

miooo0o commented on src/test/kotlin/techcourse/herobeans/client/StripeClientTest.kt:
```diff
@@ -0,0 +1,54 @@
+        "spring.jpa.hibernate.ddl-auto=none",
+    ],
+)
+@ActiveProfiles("test")
```
that's true. I add new tests applied your advice. thank you for let me know.
do you think I have to keep this test also in ours?

miooo0o committed:
Merge remote-tracking branch 'origin/dev' into feat/order-payment-flow

# Conflicts:
#	src/main/resources/application.properties

miooo0o committed:
chore: merge dev to this branch, cleanup, add new properties

miooo0o committed:
chore: disable strip client test

pebencze reviewed (approved):
Hey Mina! I think it is good to be merged now. I have left some comments that I think still need to be addressed in a future PR. I will merge now.
